system/jhtree/jhtree.cpp

```diff
@@ -1447,33 +1447,36 @@ CDiskKeyIndex::CDiskKeyIndex(unsigned _iD, IFileIO *_io, const char *_name, bool
 const CJHTreeNode *CDiskKeyIndex::loadNode(cycle_t * fetchCycles, offset_t pos, CLoadNodeCacheState & readState) const
 {
     nodesLoaded++;
-    IFileIO * useIO = readState.bufferedIO ? readState.bufferedIO.get() : io.get();
+
     unsigned nodeSize = keyHdr->getNodeSize();
 
-    //Use alloca() to allocate a buffer on the stack if the node size is small enough.
-    //Often the data could be read directly from the input files's buffer and not even be copied.
-    //Should we have a io->peek(pos, size) which returns a pointer if supported?
-    constexpr const size_t maxStackSize = 8192; // Default node size
-    MemoryAttr ma;
-    char *nodeData;
-    if (nodeSize <= maxStackSize)
-        nodeData = (char *) alloca(nodeSize);
-    else
-        nodeData = (char *) ma.allocate(nodeSize);
-    assertex(nodeData);
+    //Future: We could potentially use a different readCache for different node types.
+    CCachedIndexRead & readCache = readState.readCache;
+    const byte * match = readCache.queryBuffer(pos, nodeSize);
+    if (match)
+        return CKeyIndex::loadNodeFromMemory(match, pos, true);
+
+    // Read at least one node, but potentially multiple nodes (e.g. if from a slow storage device)
+    size32_t readSize = readState.preferredReadSize;
+    if (readSize < nodeSize)
+        readSize = nodeSize;
+
+    // Align the read position to a multiple of the readSize, and get a buffer from the read cache (reused by subsequent calls)
+    offset_t posMask = (offset_t)(readSize - 1);
+    offset_t alignedPos = (pos & ~ posMask);
+    byte * buffer = readCache.getBufferForUpdate(alignedPos, readSize);
```

Comment:
Maybe it could never happen, but if alignedPos < pos then shouldn't readSize be checked to ensure it covers pos + original readSize ?

Reply:
ok